pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -333,40 +343,91 @@ public Object visit(ASTAnnotationTypeDeclaration node, Object data) {
         return super.visit(node, data);
     }
 
-    @Override
-    public Object visit(ASTName node, Object data) {
-        // TODO: handle cases where static fields are accessed in a fully qualified way
-        //       make sure it handles same name classes and packages
-        // TODO: handle generic static field cases
+    /**
+     * Set's the node's type to the found Class in the node's name (if there is a class to be found).
+     *
+     * @param node
+     * @return The index in the array produced by splitting the node's name by '.', which is not part of the
+     * class name found. Example: com.package.SomeClass.staicField.otherField, return would be 3
+     */
+    private int searchNodeNameForClass(TypeNode node) {
+        // this is the index from which field/method names start in the dotSplitImage array
+        int startIndex = node.getImage().split("\\.").length;
+
+        // tries to find a class in the node's image by omitting the parts after each '.', example:
+        // First try: com.package.SomeClass.staticField.otherField
+        // Second try: com.package.SomeClass.staticField
+        // Third try: com.package.SomeClass <- found a class!
+        for (String reducedImage = node.getImage();;) {
+            populateType(node, reducedImage);
+            if (node.getType() != null) {
+                break; // we found a class!
+            }
 
-        Class<?> accessingClass = getEnclosingTypeDeclarationClass(node);
+            // update the start index, so that code below knows where to start in the dotSplitImage array
+            --startIndex;
 
-        String[] dotSplitImage = node.getImage().split("\\.");
-        ASTArguments astArguments = getSuffixMethodArgs(node);
-        ASTArgumentList astArgumentList = null;
-        int methodArgsArity = 0;
+            int lastDotIndex = reducedImage.lastIndexOf('.');
 
-        if (astArguments != null) {
-            astArgumentList = astArguments.getFirstChildOfType(ASTArgumentList.class);
+            if (lastDotIndex != -1) {
+                reducedImage = reducedImage.substring(0, lastDotIndex);
+            } else {
+                break; // there is no class
+            }
         }
 
-        if (astArgumentList != null) {
-            methodArgsArity = astArgumentList.jjtGetNumChildren();
+        return startIndex;
+    }
+
+    private ASTArgumentList getArgumentList(ASTArguments args) {
+        if (args != null) {
+            super.visit(args, null);
```

Comment:
it doesn't seem right to do this here.... why a getter triggers node visitation? Aren't we visiting twice these nodes by doing this?